src/settable.jl

```diff
@@ -4,32 +4,72 @@ macro settable(ex)
     esc(settable(ex))
 end
 
-function default_constructor_dict(dtype::Dict)
-    fieldsymbols = map(first, dtype[:fields])
-    body = splitdef(first(dtype[:constructors]))[:body]
-    Dict(:params => dtype[:params],
-    :body => body,
-    :name => dtype[:name],
-    :whereparams => dtype[:params],
-    :args => map(combinefield, dtype[:fields]),
-    :kwargs => [],
-    # :rtype => dtype[:name]
-    )
+argsymbol(s::Symbol) = s
+
+function argsymbol(ex::Expr)::Symbol
+    if isexpr(ex, :(::))
+        return argsymbol(ex.args[1])
+    elseif isexpr(ex, :kw)
+        return argsymbol(ex.args[1])
+    else
+        error("Unsupported expression: $(ex)")
+    end
 end
-function default_constructor(dtype::Dict)
-    combinedef(default_constructor_dict(dtype))
+
+function has_posonly_constructor(dtype::Dict)
+    fields = map(first, dtype[:fields])
+    for constructor in dtype[:constructors]
+        args = map(argsymbol, splitdef(constructor)[:args])
+        if args == fields
+            return true
+        end
+    end
+    return false
 end
+
+function posonly_constructor_dict(dtype::Dict)
+    fields = map(first, dtype[:fields])
+    for constructor in dtype[:constructors]
+        def = splitdef(constructor)
+        args = map(argsymbol, def[:args])
+        kwargs = map(argsymbol, def[:kwargs])
+        if fields[1:length(args)] == args &&
+                Set(fields[length(args)+1:end]) <= Set(kwargs)
+            newargs = copy(def[:args])
+            newkwargs = []
+            for a in def[:kwargs]
+                if argsymbol(a) in fields
+                    push!(newargs, a)
```

Comment:
Maybe we should strip default values from `newargs`. Otherwise, we are implicitly adding mutliple constructors,
e.g. `T(a,b=2) = ...` adds two constructors. These implicit constructors pollute the method table (no big deal), but might also overwrite other existing constructors and lead to subtle bugs.

Reply:
```julia
struct ManyConstructors
    a
    b
    ManyConstructors(a) = new(a,1)
    ManyConstructors(a;b=2) = new(a,b)
```

Reply:
Thanks! Good catch!

Reply:
Ah sorry that example was not good. 
```julia
struct ManyConstructors
    a
    b
    ManyConstructors(a) = new(a,1)
    ManyConstructors(a;b=2) = new(a,b)
end
```
Already overwrites the first constructor without `@settable`.
```julia
struct ManyConstructors
           a
           b
           ManyConstructors(a) = new(a,1)
           ManyConstructors(;a=1,b=2) = new(a,b)
       end
```
is a better example.